src/frontend/src/optimizer/plan_node/generic/agg.rs

```diff
@@ -101,6 +207,26 @@ impl<PlanRef: GenericPlanRef> Agg<PlanRef> {
         self.ctx().session_ctx().config().get_enable_two_phase_agg()
     }
 
+    pub(crate) fn can_two_phase_agg(&self) -> bool {
+        self.two_phase_agg_enabled()
+            && !self.agg_calls.is_empty()
+            && self.agg_calls.iter().all(|call| {
+                let order_ok = matches!(call.agg_kind, agg_kinds::result_unaffected_by_order_by!())
+                    || call.order_by.is_empty();
+                let distinct_ok =
+                    matches!(call.agg_kind, agg_kinds::result_unaffected_by_distinct!())
+                        || !call.distinct;
+                order_ok
+                    && distinct_ok
+                    && !matches!(call.agg_kind, agg_kinds::simply_cannot_two_phase!())
```

Comment:
minor: prefer that we also assign a separate variable to this, similar to `order_ok` and `distinct_ok`.

```suggestion
                let agg_kind_ok = !matches!(call.agg_kind, agg_kinds::simply_cannot_two_phase!());
                order_ok
                    && distinct_ok
                    && agg_kind_ok
```